npm/podController.go

```diff
@@ -472,47 +472,36 @@ func (c *podController) syncAddAndUpdatePod(newPodObj *corev1.Pod) error {
-	addToIPSets := []string{}
 	// compare cached NPMPod's IP address against newPod's IP address
-	if cachedPobObjIpChanged {
-		metrics.SendErrorLogAndMetric(util.PodID, "[syncAddAndUpdatePod] Info: Unexpected state. Pod (Namespace:%s, Name:%s, uid:%s, has cachedPodIp:%s which is different from PodIp:%s",
-			newPodObjNs, newPodObj.Name, cachedNpmPodObj.PodUID, cachedNpmPodObj.PodIP, newPodObj.Status.PodIP)
-		// cached PodIP needs to be cleaned up from all the cached labels
-		deleteFromIPSets = util.GetIPSetListFromLabels(cachedNpmPodObj.Labels)
-
-		// Assume that the pod IP will be released when pod moves to succeeded or failed state.
-		// If the pod transitions back to an active state, then add operation will re-establish the updated pod info.
-		// new PodIP needs to be added to all newLabels
-		addToIPSets = util.GetIPSetListFromLabels(newPodObj.Labels)
-
-		klog.Infof("Deleting pod %s %s from ipset %s and adding pod %s to ipset %s",
-			cachedNpmPodObj.PodUID, cachedNpmPodObj.PodIP, cachedNpmPodObj.Namespace, newPodObj.Status.PodIP, newPodObjNs,
-		)
-
-		// Delete the pod from its namespace's ipset.
-		if err = ipsMgr.DeleteFromSet(util.GetNSNameWithPrefix(cachedNpmPodObj.Namespace), cachedNpmPodObj.PodIP, podKey); err != nil {
-			return fmt.Errorf("[syncAddAndUpdatePod] Error: failed to delete pod from namespace ipset with err: %v", err)
+	// NPM should clean up existing references of cached pod obj and its IP.
+	// then, re-add new pod obj.
+	if cachedNpmPodObj.PodIP != newPodObj.Status.PodIP {
+		metrics.SendErrorLogAndMetric(util.PodID, "[syncAddAndUpdatePod] Info: Unexpected state. Pod (Namespace:%s, Name:%s, uid:%s, newUid:%s) , has cachedPodIp:%s which is different from PodIp:%s",
+			newPodObjNs, newPodObj.Name, cachedNpmPodObj.PodUID, string(newPodObj.UID), cachedNpmPodObj.PodIP, newPodObj.Status.PodIP)
+
+		klog.Infof("Deleting cached Pod with key:%s first due to IP Mistmatch", podKey)
+		if err = c.cleanUpDeletedPod(cachedNpmPodObj.getPodObjFromNpmPodObj()); err != nil {
+			return err
 		}
-	} else {
-		// the IP addresses of the cached npmPod and newPodObj is the same
-		// If no change in labels, then GetIPSetListCompareLabels will return empty list.
-		// Otherwise it returns list of deleted PodIP from cached pod's labels and list of added PodIp from new pod's labels
-		addToIPSets, deleteFromIPSets = util.GetIPSetListCompareLabels(cachedNpmPodObj.Labels, newPodObj.Labels)
+
+		klog.Infof("Adding back Pod with key:%s after IP Mistmatch", podKey)
+		if err = c.syncAddedPod(newPodObj); err != nil {
+			return err
+		}
+
+		return nil
 	}
 
+	// the IP addresses of the cached npmPod and newPodObj is the same
+	// If no change in labels, then GetIPSetListCompareLabels will return empty list.
+	// Otherwise it returns list of deleted PodIP from cached pod's labels and list of added PodIp from new pod's labels
+	addToIPSets, deleteFromIPSets := util.GetIPSetListCompareLabels(cachedNpmPodObj.Labels, newPodObj.Labels)
```

Comment:
From here, do we deal with the first case? If so, can we keep using cachedNpmPodObj and update the  cachedNpmPodObj if needed? Are there specific reasons to create newNpmPod() and update cache, etc?

Reply:
the reason we are creating new obj is because we have fields like Phase, PodIPs, Rv etc which needs to be updated to reflect the latest newPodobj ... we can add them individually too if it makes sense

Reply:
Good improvement, adding this change